src/pepr/operator/controllers/keycloak/clients/client-credentials.ts

```diff
@@ -0,0 +1,150 @@
+  await throwErrorIfNeeded(response);
+  cachedToken = response.data.access_token;
+  return cachedToken;
+}
+
+export async function credentialsCreateOrUpdate(client: Partial<Client>) {
+  log.info(`credentialsCreateOrUpdate: creating/updating client ${JSON.stringify(client)}`);
+  const existingClient = await credentialsGet(client);
+  if (existingClient) {
+    return credentialsUpdate(client);
+  } else {
+    return credentialsCreate(client);
+  }
+}
+
+export async function credentialsCreate(client: Partial<Client>) {
+  log.info(`credentialsCreate: creating client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const response = await fetch(clientsAdminUrl, {
+    method: "POST",
+    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
+    body: JSON.stringify(client),
+  });
+  await throwErrorIfNeeded(response, () => {
+    resetCachedToken();
+  });
+  return credentialsGet(client);
+}
+
+export async function credentialsGet(client: Partial<Client>) {
+  log.info(`credentialsGet: retrieving client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const url = `${clientsAdminUrl}?clientId=${encodeURIComponent(client.clientId!)}`;
+  // There's no Client GET REST endpoint that obtains a client based on client_id (the logical client name, like uds-operator).
+  // All Admin REST endpoints for client operator on the database Client ID, which is a UUID. The only interface that allows to
+  // obtain the Client using the client_id is the collection interface which returns a singular collection with
+  // the Client in it.
+  const response = await fetch<ClientWithId[]>(url, {
+    method: "GET",
+    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
+  });
+  await throwErrorIfNeeded(response, () => {
+    resetCachedToken();
+  });
+  return response.data[0];
+}
+
+export async function credentialsUpdate(client: Partial<Client>) {
+  log.info(`credentialsUpdate: updating client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const existing = await credentialsGet(client);
+  if (!existing || !existing.id) {
+    throw new Error(`Failed to retrieve existing client, ${client.clientId}`);
+  }
+  const url = `${clientsAdminUrl}/${encodeURIComponent(existing.id)}`;
+  const response = await fetch(url, {
+    method: "PUT",
+    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
+    body: JSON.stringify(client),
+  });
+  await throwErrorIfNeeded(response, () => {
+    resetCachedToken();
+  });
+  return credentialsGet(client);
+}
+
+export async function credentialsDelete(client: Partial<Client>) {
+  log.info(`credentialsDelete: deleting client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const existing = await credentialsGet(client);
+  const url = `${clientsAdminUrl}/${encodeURIComponent(existing.id!)}`;
```

Comment:
```suggestion
  if (!existing || !existing.id) {
    throw new Error(`Failed to retrieve existing client, ${client.clientId}`);
  }
  const url = `${clientsAdminUrl}/${encodeURIComponent(existing.id)}`;
```

Same change as for the update func above.